add `.` and `../` completions

meganrogge opened the pull request:
fixes https://github.com/microsoft/vscode/issues/234351

meganrogge committed:
fix #234351

meganrogge committed:
tweak

Tyriar commented:
Testing this out now. The `../` here should re-trigger the completions on Windows only, just as `..\` does. Right now it just continues the filter the items:

![image](https://github.com/user-attachments/assets/6140f2ab-7efa-4ccf-bff7-cad7aebbc31b)

Tyriar commented:
I'm not seeing `.` show up on Windows. This is what I see:

![image](https://github.com/user-attachments/assets/a9b3ecfa-4079-40e2-bfa4-af2e9896af42)

Bottom of the list:

![image](https://github.com/user-attachments/assets/ceec7196-b47a-4291-9441-23179ee2723c)

meganrogge committed:
consolidate

Tyriar commented:
When you complete `cd ..\` I see this:

![image](https://github.com/user-attachments/assets/2fec6cfb-f4c1-45fe-bff0-57c0164015ee)

Notice the odd filtering of the very last item `..\`. It should show up first in the list and be fully blue? This would make it work really well with `"terminal.integrated.suggest.runOnEnter": "exactMatchIgnoreExtension"` and `exactMatch`.

meganrogge committed:
move . and .. to top

Tyriar commented on src/vs/workbench/contrib/terminalContrib/suggest/browser/terminalCompletionService.ts:
```diff
@@ -200,6 +200,43 @@ export class TerminalCompletionService extends Disposable implements ITerminalCo
+			if (!lastWord.trim()) {
+				resourceCompletions.push({
+					label: '.',
+					provider: 'builtin',
```
Shouldn't this include the actual provider since we want to know what extension(s) are requesting those resources.

meganrogge commented:
> When you complete `cd ..\` I see this:
> 
> ![image](https://private-user-images.githubusercontent.com/2193314/402682675-2fec6cfb-f4c1-45fe-bff0-57c0164015ee.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzY3OTk1NzUsIm5iZiI6MTczNjc5OTI3NSwicGF0aCI6Ii8yMTkzMzE0LzQwMjY4MjY3NS0yZmVjNmNmYi1mNGMxLTQ1ZmUtYmZmMC01N2MwMTY0MDE1ZWUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDExMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAxMTNUMjAxNDM1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjFmOTFhNmE2NTgwZDk5MDU5YmU2NTFlOWZjZjIxNjIzMzIxYWM5YjhlMWY0ZDQ4ZjVjNzU4ODQ5YjgyZmFjMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.T7Qa0ZvY-SowMMovadFRxRbEY1UtndosmIUS6u_NmCA)
> 
> Notice the odd filtering of the very last item `..\`. It should show up first in the list and be fully blue? This would make it work really well with `"terminal.integrated.suggest.runOnEnter": "exactMatchIgnoreExtension"` and `exactMatch`.

that's because looks like it has the wrong replacement index?

Tyriar commented on src/vs/workbench/contrib/terminalContrib/suggest/browser/terminalCompletionService.ts:
```diff
@@ -200,6 +200,43 @@ export class TerminalCompletionService extends Disposable implements ITerminalCo
+					label: '..' + resourceRequestConfig.pathSeparator,
+					provider: 'builtin',
+					kind: TerminalCompletionItemKind.Folder,
+					detail: 'Parent folder',
```
Let's add the actual folders here since we only get this far when `cwd` is known

Tyriar committed:
Push WIP for fixing edge cases

meganrogge commented on src/vs/workbench/contrib/terminalContrib/suggest/browser/terminalCompletionService.ts:
```diff
@@ -252,23 +290,31 @@ export class TerminalCompletionService extends Disposable implements ITerminalCo
-				}
-			}
+			// TODO: This doesn't do the windows path equivalence check
+			// if (!lastWord.startsWith('.' + resourceRequestConfig.pathSeparator) && !lastWord.startsWith('..' + resourceRequestConfig.pathSeparator)) {
```
I think we need this still?

Tyriar commented:
Some test cases to work through when we meet:

Test cases:

- Windows
    - `.|` should show `.` (top) and `..\`
    - `..\|` should show `..\` (top) and `..\..\`
    - `../|` should show `..\` (top) and `..\..\`
    - `..\` should show directories of `..\` (top)
    - `../..\` should show `..\..\` (top) and `..\..\..\`
- Mac
    - When directories are completed, the completed dir should always show at the top result (eg. `.`, `..`, `../`, `../../`, `./src/`, `../xterm.js`, etc.)
- Both
    - Resolve actual folder name as detail for `.`, `..\`, `~`, etc.
    - Filtering on front end
        - Triggering at `./sr|` should show the exact same results as when it's triggered at `./` and `sr` is typed
        - Triggering at `./src/cm|` should find `./src/common`
    - Folders should be resolved even if the trailing `/` is missing
        - `./src` should show `./src/` (top) and `./src/<other_dirs>`

Discussion:

- `../` Scoring
    - Since this is easy to type and probably a low chance of being completed, should this go to the bottom unless it's an exact match?
        - eg. `./src/` shows `./src/` then `./src/../`, taking up valuable real estate

- Define behavior for when `../` shows up exactly
    - Should we show when another folder is typed?
        - Eg. `./src/` showing `./src/..`?
            - If so, should that completion be smarter and "complete" to `./`?
        - Should we just not show `<prefix>../` unless it matches `/(\.\.\/)+/`?

meganrogge committed:
update test expectations, get tests to pass

meganrogge committed:
use provider, not builtin

meganrogge committed:
provide paths as details

meganrogge committed:
add test case for and fix bug

meganrogge committed:
fix tests, some bugs

meganrogge committed:
Refactor, clean up the code

Co-authored-by: Daniel Imms <Tyriar@users.noreply.github.com>

meganrogge committed:
further simplify

meganrogge committed:
add a comment

meganrogge committed:
better name

meganrogge committed:
sort by label alphabetically

Tyriar commented on src/vs/workbench/services/suggest/browser/simpleCompletionModel.ts:
```diff
@@ -207,6 +207,7 @@ export class SimpleCompletionModel {
 				}
 				// Then by file extension length ascending
 				score = a.fileExtLow.length - b.fileExtLow.length;
+				score = a.completion.label.localeCompare(b.completion.label);
```
Ah that explains the alphabetical ordering. This change clobbers the extension length part so make sure that doesn't break.

Tyriar commented on src/vs/workbench/contrib/terminalContrib/suggest/test/browser/terminalCompletionService.test.ts:
```diff
@@ -113,15 +161,40 @@ suite('TerminalCompletionService', () => {
+			function assertCompletions(actual: ITerminalCompletion[] | undefined, expected: IAssertionTerminalCompletion[], expectedConfig: IAssertionCommandLineConfig) {
+				assert.deepStrictEqual(
+					actual?.map(e => ({
+						label: e.label,
```
Let's replace `/` with `pathSeparator` here? That'll make tests look even nicer.

meganrogge committed:
Add comment

meganrogge committed:
refactor tests

meganrogge committed:
move score part into conditional to not override it

meganrogge committed:
add test for alphabetical order

meganrogge committed:
Add more tests

meganrogge committed:
add more tests

meganrogge committed:
rm commas

meganrogge committed:
rename

Tyriar commented:
To fix this we could pass the windows thing into `resourceRequestConfig`, maybe as `secondaryPathSeparators` or something like that to make it more generic?

Tyriar commented:
We need to handle the Windows primary paths in the test:

<img width="457" alt="Screenshot 2025-01-14 at 3 30 09 pm" src="https://github.com/user-attachments/assets/f657ca18-6597-4329-b1b4-000b1bb3c359" />

Tyriar commented on src/vs/workbench/contrib/terminalContrib/suggest/test/browser/terminalCompletionService.test.ts:
```diff
@@ -67,129 +95,247 @@ suite('TerminalCompletionService', () => {
+			];
+			const result = await terminalCompletionService.resolveResources(resourceRequestConfig, '/usr/', 5, provider);
+
+			assertCompletions(result, [
+				{ label: '/usr/' },
+				{ label: './usr/' },
+				{ label: './home/' },
+				{ label: '/usr/../' }
+			], { replacementIndex: 0, replacementLength: 5 });
```
This should actually not include the `.` ones here since `/` is how you signal an absolute path when not on Windows.

Tyriar commented on src/vs/workbench/contrib/terminalContrib/suggest/test/browser/terminalCompletionService.test.ts:
```diff
@@ -67,129 +95,247 @@ suite('TerminalCompletionService', () => {
+			], { replacementIndex: 0, replacementLength: 5 });
+		});
+
+		test('./folder| Case sensitivity should resolve correctly on Mac/Unix', async () => {
```
Do we need to test this only on !windows and have the opposite test for Windows?

Tyriar commented on src/vs/workbench/contrib/terminalContrib/suggest/test/browser/terminalCompletionService.test.ts:
```diff
@@ -67,129 +95,247 @@ suite('TerminalCompletionService', () => {
+			assert(result);
+			// includes the 1000 folders + ./ and ./../
+			assert.strictEqual(result?.length, 1002);
+			assert.strictEqual(result[0].label, './');
```
assert `result.at(-1)` too?

Tyriar commented on src/vs/workbench/contrib/terminalContrib/suggest/test/browser/terminalCompletionService.test.ts:
```diff
@@ -67,129 +95,247 @@ suite('TerminalCompletionService', () => {
-			}]);
-		});
-		test('cd ./f|', async () => {
+			const result = await terminalCompletionService.resolveResources(resourceRequestConfig, './h', 3, provider);
```
We should make it so we pass in the Windows state here, since this is called on the frontend. Maybe via something like `resourceRequestConfig.secondaryPathSeparators: string[]`?

meganrogge replied:
how about replacing all / or \ in the prefix with the pathSeparator?

meganrogge replied:
done

meganrogge replied:
hm, i'm thinking that's less efficient, so will do the secondaryPathSeparators idea or similar instead

meganrogge committed:
also assert last item

meganrogge committed:
add isWindows and opposite test case

meganrogge committed:
normalize prefix

meganrogge committed:
pass in should normalize prefix

meganrogge committed:
check if absolute path before adding children

Tyriar commented on src/vs/workbench/contrib/terminalContrib/suggest/browser/terminalCompletionService.ts:
```diff
@@ -248,47 +248,52 @@ export class TerminalCompletionService extends Disposable implements ITerminalCo
 			});
 		}
 
+		// Handle absolute paths differently to avoid adding `./` prefixes
+		const isAbsolutePath = lastWord.startsWith(resourceRequestConfig.pathSeparator) && lastWord.endsWith(resourceRequestConfig.pathSeparator);
```
Absolute paths look differently on Windows: `/^[A-Z]:\/

meganrogge committed:
fix a test

meganrogge committed:
rm a test

meganrogge committed:
alter test

meganrogge committed:
add test for absolute paths on windows

Tyriar commented on src/vs/workbench/contrib/terminalContrib/suggest/browser/terminalCompletionService.ts:
```diff
@@ -183,6 +186,10 @@ export class TerminalCompletionService extends Disposable implements ITerminalCo
 	}
 
 	async resolveResources(resourceRequestConfig: TerminalResourceRequestConfig, promptValue: string, cursorPosition: number, provider: string): Promise<ITerminalCompletion[] | undefined> {
+		if (resourceRequestConfig.shouldNormalizePrefix) {
+			// for tests, make sure the right path separator is used
+			promptValue = promptValue.replaceAll(/[\\/]/g, resourceRequestConfig.pathSeparator);
+		}
```
We want the flag here to eliminate the rest of the `isWindows` checks so we can control this for the unit tests.

Most of Windows shells I expect to work the same, but a special case we need to consider here is that for git bash we want it to only support `/` and use its special style of absolute paths that end up mapping to real Windows paths:

![image](https://github.com/user-attachments/assets/4ecfa140-577e-4f34-b948-37f388240a87)

We may need to do something similar for `/mnt/` in WSL shells.

meganrogge replied:
Oh, ok. Will do.

meganrogge replied:
I think we still want this `isWindows` check?

https://github.com/microsoft/vscode/blob/0f33b3dacea552905c6d8a72affc58c709d8e76f/src/vs/workbench/contrib/terminalContrib/suggest/browser/terminalCompletionService.ts#L254-L257

Tyriar commented on src/vs/workbench/contrib/terminalContrib/suggest/browser/terminalCompletionService.ts:
```diff
@@ -197,46 +204,117 @@ export class TerminalCompletionService extends Disposable implements ITerminalCo
+
+		// Add parent directory to the bottom of the list because it's not as useful as other suggestions
+		//
+		// For example:
+		// - `|` -> `../`
+		// - `./src/|` -> `./src/../`
+		//
+		// On Windows, the path seprators are normalized to `\`:
+		// - `./src/|` -> `.\src\..\`
+		if (foldersRequested) {
+			const parentDir = URI.joinPath(cwd, '..' + resourceRequestConfig.pathSeparator);
 			resourceCompletions.push({
```
Should we block parent dir for `isAbsolutePath` too? `/../` is an invalid suggestion and `/usr/../` is a strange suggestion

meganrogge replied:
Just seeing this, yep, makes sense, will do

Tyriar committed:
Fix filling of children dirs on Windows, fix unit tests on Windows

Tyriar committed:
Remove most recent completion feature

This was used for pwsh completions to ensure the last completed completion
always shows up so runOnEnter exactMatch works well. Path completions are
entirely handled in completion service now so we shouldn't need this special
behavior.

Tyriar committed:
Show absolute path in completion details

meganrogge committed:
fix test expectations

meganrogge committed:
fix test expectations

meganrogge committed:
don't provide parent dir for absolute path

meganrogge committed:
set as var

meganrogge committed:
use path sep in else

meganrogge committed:
add comment, todo
